### DISCUSSION TITLE
Rubi support

### DISCUSSION THREAD

**Discussion by aravindh-krishnamoorthy** [Mathics3/Mathics3-Rubi#1] [General]
This is a general thread about current Rubi support.

# Current status

> [!NOTE]
> A very small subset of Rubi is usable. See below for instructions.

- [ ] `1 Algebraic functions`
    - [ ] `1.1 Binomial products`
        - [ ] `1.1.1 Linear`
            - [X] `1.1.1.1  (a+b x)^m`
            - [ ] `1.1.1.2 (a+b x)^m (c+d x)^n`   (⚠️Loaded but hangs. Currently being debugged.)
            - [ ] `...`
        - [ ] `...`
   - [ ]   `...`

There are nine more sections besides  `1 Algebraic functions`. So, a long way to go!

# Usage instructions

```bash
$ git clone -b rubi git@github.com:aravindh-krishnamoorthy/mathics-core.git
$ cd mathics-core/mathics/packages/Rubi
$ mathics

Mathics 7.0.1dev0
on CPython 3.11.10 | packaged by conda-forge | (main, Oct 16 2024, 01:27:36) [GCC 13.3.0]
using SymPy 1.13.3, mpmath 1.3.0, numpy 2.1.3, cython Not installed

...
```
```mathematica
In[1]:= << Rubi.m
Loading Rubi 4.17.3.0 will take a minute or two. In the future this will take less than a second.
Loading /home/ark/git/mathics/mathics-core/mathics/packages/Rubi/IntegrationUtilityFunctions.m...

Loading /home/ark/git/mathics/mathics-core/mathics/packages/Rubi/IntegrationRules/1 Algebraic functions/1.1 Binomial products/1.1.1 Linear/1.1.1.1 (a+b x)^m.m...

Loading /home/ark/git/mathics/mathics-core/mathics/packages/Rubi/IntegrationRules/1 Algebraic functions/1.1 Binomial products/1.1.1 Linear/1.1.1.2 (a+b x)^m (c+d x)^n.m...

Loading /home/ark/git/mathics/mathics-core/mathics/packages/Rubi/IntegrationRules/1 Algebraic functions/1.1 Binomial products/1.1.1 Linear/1.1.1.3 (a+b x)^m (c+d x)^n (e+f x)^p.m...

Loading /home/ark/git/mathics/mathics-core/mathics/packages/Rubi/IntegrationRules/1 Algebraic functions/1.1 Binomial products/1.1.1 Linear/1.1.1.4 (a+b x)^m (c+d x)^n (e+f x)^p (g+h x)^q.m...

Out[1]= None

In[2]:= Int[(b x + a)^m, x]
Out[2]= (a + b x) ^ (1 + m) / (b (1 + m))

In[3]:= Int[x, x]
Out[3]= x ^ 2 / 2

In[4]:= Int[(a + x)^2, x]
Out[4]= (a + x) ^ 3 / 3

In[5]:= Int[x^3, x]
Out[5]= x ^ 4 / 4

In[6]:= Int[(a + x)^m, x]
Out[6]= (a + x) ^ (1 + m) / (1 + m)

In[7]:= Int[(a + b x)^m (c + d x)^n, x]
(* Doesn't complete within reasonable time *)
Out[7]= $Aborted
```

**Comment by rocky** [comment]
Thanks for undertaking this!

For what it's worth, when I run `TraceEvaluation[]` on the expression that does not complete, I am seeing: 

```
In[20]:= TraceEvaluation[Int[(a + b x)^m (c + d x)^n, x]]
    Evaluating: Rubi`Int[System`Times[System`Power[System`Plus[Global`a, System`Times[Global`b, Global`x]], Global`m], System`Power[System`Plus[Global`c, System`Times[Global`d, Global`x]], Global`n]], Global`x]
      Evaluating: Rubi`Int
      Evaluating: System`Times[System`Power[System`Plus[Global`a, System`Times[Global`b, Global`x]], Global`m], System`Power[System`Plus[Global`c, System`Times[Global`d, Global`x]], Global`n]]
        Evaluating: System`Times
        Evaluating: System`Power[System`Plus[Global`a, System`Times[Global`b, Global`x]], Global`m]
          Evaluating: System`Power
          Evaluating: System`Plus[Global`a, System`Times[Global`b, Global`x]]
            Evaluating: System`Plus
            Evaluating: Global`a
            Evaluating: System`Times[Global`b, Global`x]
              Evaluating: System`Times
              Evaluating: Global`b
              Evaluating: Global`x
          Evaluating: Global`m
          Evaluating: System`Plus[Global`a, System`Times[Global`b, Global`x]]
            Evaluating: System`Plus
            Evaluating: Global`a
            Evaluating: System`Times[Global`b, Global`x]
              Evaluating: System`Times
              Evaluating: Global`b
              Evaluating: Global`x
          Evaluating: Global`m
        Evaluating: System`Power[System`Plus[Global`c, System`Times[Global`d, Global`x]], Global`n]
          Evaluating: System`Power
          Evaluating: System`Plus[Global`c, System`Times[Global`d, Global`x]]
            Evaluating: System`Plus
            Evaluating: Global`c
            Evaluating: System`Times[Global`d, Global`x]
              Evaluating: System`Times
              Evaluating: Global`d
              Evaluating: Global`x
          Evaluating: Global`n
          Evaluating: System`Plus[Global`c, System`Times[Global`d, Global`x]]
            Evaluating: System`Plus
            Evaluating: Global`c
            Evaluating: System`Times[Global`d, Global`x]
              Evaluating: System`Times
              Evaluating: Global`d
              Evaluating: Global`x
          Evaluating: Global`n
      Evaluating: Global`x
      Evaluating: System`Default[System`Times, 2]
        Evaluating: System`Default
        Evaluating: System`Times
      Evaluating: System`Default[System`Times, 2]
        Evaluating: System`Default
        Evaluating: System`Times
      Evaluating: System`Default[System`Times, 2]
        Evaluating: System`Default
  loops forever ...
```

And I see: 

``` 
Evaluating: System`Plus[Global`c, System`Times[Global`d, Global`x]]

```

two times.

**Reply by rocky** [reply]
I have been working on a debugger for Mathics3 and as part of that a better version of TraceEvaluation. There definitely is an infinite or long-running loop we have.: 

![image](https://github.com/user-attachments/assets/6524149a-30f5-4baf-9f12-33ea1ef88f8c)

I will use real examples like this to improve the debugging experience.

**Reply by rocky** [reply]
The setup to get this was: 

![image](https://github.com/user-attachments/assets/9af9b1c9-1f10-4d0c-9f54-7a24dbab4bf1)

**Reply by rocky** [reply]
And more info... 

![image](https://github.com/user-attachments/assets/d7522cc5-e26c-4730-94cb-abc203c84416)

So basically we are stuck in trying to `rewrite_apply_eval_step()` with  `Default[Times, 2]`,  evaluating it, and getting 1 back and then going back to `rewrite_apply_eval_step()`. 

At some point, I'll fix things so that you can enter "Default" instead of "System`Default".

**Reply by aravindh-krishnamoorthy** [reply]
Ok, this is strange. The example should quickly return the expression unevaluated or an expression based on hypergeomeric functions, depending on which Rubi rules are loaded.

**Reply by aravindh-krishnamoorthy** [reply]
Based on a first quick look, this seems like a performance issue with handling patterns with default values in Python functions `match` and `match_expression_with_one_identity` (file: `pattern.py`). I'll look a bit deeper into these and see if my hunch checks out... 

ALso, if we let it run for a long time with just `1.1.1.1` to `1.1.1.4` loaded, the tracing for `TraceEvaluation[Int[(a + b x)^m (c + d x)^n, x]]` stops after a while! This might be why the function does not return...

**Reply by aravindh-krishnamoorthy** [reply]
@rocky Unfortunately, I think the problems are more basic:

```mathematica
In[1]:= f[(a_. + b_. x_Symbol)^n_.] = 1
Out[1]= 1

# Ok
In[2]:= f[2 x]
Out[2]= 1

# Wrong!
In[3]:= f[1 + 2 x]
Out[3]= f[1 + 2 x]

# Wrong!
In[4]:= f[(1 + 2 x)^1]
Out[4]= f[1 + 2 x]

# Ok!
In[5]:= f[(0 + 2 x)^1]
Out[5]= 1

In[6]:=
```

**Reply by rocky** [reply]
@aravindh-krishnamoorthy It took me a while to understand the WMA-specific pattern-matching gobbledygook above.  Although you did a good job isolating the problem somewhat, for me, a WMA novice, to begin to be able to fix the problems, it would help to break this down more into simpler, or even better the the _simplest_ cases. 

Let me describe my understanding of the situation here in the hope that in the future we can break down problems into the kinds of things that someone who doesn't understand the WMA language.  The more things are narrowed, the easier it is to address problems.

For context, I gather this was motivated by the `Default[]` function <strike>that Rubi seems to define</strike>, a Builtin Function of Mathics3. (In tracing, we saw `Default[Times, 2]]`.) 

Here, what we have is a pattern-matching problem.  These are more easily worked on using the `MatchQ[]` function. So restating this:

```In[1]:= MatchQ[2 x, (a_. + b_. x_Symbol)^n_.]
Out[1]= True

In[2]:= MatchQ[1 + 2 x^1, (a_. + b_. x_Symbol)^n_.]
Out[2]= False (* Should be True *)

In[3]:= MatchQ[(0 + 2 x)^1, (a_. + b_. x_Symbol)^n_. ]
Out[3]= True
```

So I gather that "+" (is this pattern _or_ ?)  is failing when _a_ is 1 but not 0.  It looks like the `^n_.` is not relevant here.  So I tried removing this and, sure enough, I got the wrong behavior:

```
In[4]:= MatchQ[0 + 2 x, (a_. + b_. x_Symbol)]
Out[4]= True

In[5]:= MatchQ[1 + 2 x, (a_. + b_. x_Symbol)]
Out[5]]= False
```

Can I go further, and remove the space (" ") in `b_. x_Symbol`? 

```
In[16]:= MatchQ[0 + x, (a_. + b_.)]
Traceback (most recent call last):
  File "/home/rocky/.pyenv/versions/3.12.7/bin/mathics", line 33, in <module>
    sys.exit(load_entry_point('Mathics3', 'console_scripts', 'mathics')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Changing to `b_`  which does work and `x_Symbol` works as well.

But at any rate, this is enough for me to start investigating the pattern-matching bugs we have. I will write up a bug report in mathics-core soon.

**Reply by mmatera** [reply]
@rocky, notice that `MatchQ[0 + 2 x, (a_. + b_. x_Symbol)]` is evaluated first to ` MatchQ[2 x, (a_. + b_. x_Symbol)]` before trying to match the pattern. Maybe you should wrap the pattern and the expression with `Hold`

**Reply by rocky** [reply]
Ok - thanks. In `f[(a_. + b_. x_Symbol)^n_.] = 1` there is an implied `Hold[]` as well?  

@aravindh-krishnamoorthy  thanks for correcting that `Default[]` is a builtin Mathics3 function.

**Comment by rocky** [comment]
@axkr You got this working for Symja, right? 

Do you have any advice on the process of getting a large package like this working? Or specific advice for debugging this package?  Thanks.

**Reply by axkr** [reply]
> Do you have any advice on the process of getting a large package like this working? Or specific advice for debugging this package?

We've taken a complete other way, which is described here:
https://github.com/axkr/symja_android_library/wiki/Porting-Rubi-Integration-rules-to-Symja

@halirutan helped in getting the Rubi rules in this special format:
- https://github.com/axkr/symja_android_library/tree/master/symja_android_library/Rubi

If you go this way, IMO you should try `RubiRules4.16.0_FullLHS.m` first. 
It contains lesser rules than `RubiRules_4.17.3.0_FullLHS.m` and worked more stable in my environment.

**Reply by rocky** [reply]
Thanks for the information. Very interesting. It will take a while to look over and fully understand.

**Reply by aravindh-krishnamoorthy** [reply]
Thank you, axkr and rocky for the inputs! My first attempt would be to understand the above issues and try and get the official Rubi running without any changes (since Rubi runs as-is in Mathematica). If it doesn't work at all, then perhaps we'd have to "port" the rules and/or change some of the conditions in "IntegrationUtilityFunctions".

**Reply by axkr** [reply]
One special thing I remember was, that I have to add the `Star` operator for the 4.17 version. But I assume this operator is already implemented in Mathics?
- https://reference.wolfram.com/language/ref/Star.html
- https://github.com/axkr/symja_android_library/commit/fb13a9e2f8a85bdfedc679901ddc4fc5c8ad24ef

**Comment by rocky** [comment]
@aravindh-krishnamoorthy we should definitely use (adapted for Python) the Integration test suite that Symja has developed. See https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-io/src/test/java/org/matheclipse/core/rubi/AlgebraicFunctions.java

@axkr One thing that confuses me about these tests is that they call `Integrate[]` rather than the Rubi `Int[]`.   Does Symja rely on Rubi to perform all integration?

**Reply by axkr** [reply]
> we should definitely use (adapted for Python) the Integration test suite that Symja  we should definitely use (adapted for Python) the integration test suite has developed.  

the test suite was also generated by @halirutan. Maybe he can modify his script for python.
Above the test there's the line-number (from the 4.16 version) of the rule which should match first for the example:
![image](https://github.com/user-attachments/assets/d298ae00-7f1d-46ec-aa6d-68c0cf5d3a53)


> @axkr One thing that confuses me about these tests is that they call `Integrate[]` rather than the Rubi `Int[]`. Does Symja rely on Rubi to perform all integration?

yes the `Integrate` function relies on Rubi plus some extra hard coded features:
- https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-core/src/main/java/org/matheclipse/core/reflection/system/Integrate.java

**Reply by aravindh-krishnamoorthy** [reply]
For now, I did this: https://github.com/aravindh-krishnamoorthy/mathics-core/tree/ccb2219c064774b0f13c50ef32276f30b1bb1876/mathics/packages/Rubi/test but a long term strategy with test numbers would be very welcome.

**Reply by axkr** [reply]
> For now, I did this: https://github.com/aravindh-krishnamoorthy/mathics-core/tree/ccb2219c064774b0f13c50ef32276f30b1bb1876/mathics/packages/Rubi/test but a long term strategy with test numbers would be very welcome.

Comparing only the strings for tests will not work for complicated expressions.

I try to use my (not very perfect version) of `PossibleZeroQ` to first subtract the result from the expected expression and see if it gets `0` or I do derive the result w.r.t. `x`  and subtract the original integrate function and see if it gets `0`

**Reply by aravindh-krishnamoorthy** [reply]
I see, thank you for the inputs, @axkr. I will keep this in mind. Right now, I'm debugging very basic things like matching default patterns, where I'm running into trouble. But, once this is set, I expect that we can begin integrating the algebraic tests.

**Reply by rocky** [reply]
> > we should definitely use (adapted for Python) the Integration test suite that Symja  we should definitely use (adapted for Python) the integration test suite has developed.
> 
> the test suite was also generated by @halirutan. Maybe he can modify his script for python. Above the test there's the line-number (from the 4.16 version) of the rule which should match first for the example: ![image](https://private-user-images.githubusercontent.com/3109793/387318275-d298ae00-7f1d-46ec-aa6d-68c0cf5d3a53.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE5NTcxNjMsIm5iZiI6MTczMTk1Njg2MywicGF0aCI6Ii8zMTA5NzkzLzM4NzMxODI3NS1kMjk4YWUwMC03ZjFkLTQ2ZWMtYWE2ZC02OGMwY2Y1ZDNhNTMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMThUMTkwNzQzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWY3NmFlMDkxZWEwNWMyYWQyNDQ0N2RlYTJiMGMwNTJlYWI3MjJjNzdhYWU2OGU0NmEzM2RjNzQ4OTk4MzFhNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.-NZv8XlZthy-hqIqVl-fauuo61tRTdOvq3kfTUtejeU)

Interesting. I guess I didn't mean "modified for Python" so much as _not_ in Java format.  I assumed this was hand generated since it is checked into git. But what format were the tests originally? If WMA, that is okay. 

What I guess I'd like happen is that unless this is already in WMA, it be put into some language-agnostic format like YAML. (YAML is slightly better than JSON because it supports adding comments.)   And then this can be in a common repository were Mathics3 could read the YAML into Python and run it that way.   

The common YAML could have a field to indicate not to run for a particular project like Mathics3 or Symja.

**Reply by axkr** [reply]
> But what format were the tests originally? If WMA, that is okay.

The tests are somehow extracted from the testsuite:
- https://rulebasedintegration.org/TestFiles/MathematicaSyntaxFiles/MathematicaSyntaxTestFiles.zip
- https://rulebasedintegration.org/testProblems.html

**Reply by rocky** [reply]
@aravindh-krishnamoorthy Since the test suite from Rubi is provided in WMA format, let's try to use that.

**Comment by halirutan** [comment]
I can give some insight into what exactly we did when testing Symja. You can test three things when you want to compare Mathics behavior to how Rubi works in Mathematica:

1. You use a test case like `Int[F^(2 + 5*x), x]` and compare your result with the result of Rubi. As @axkr already mentioned, this can be tricky because even in medium-sized problems, you run into issues where equivalent results cannot be verified with a simple string comparison.
2. You preprocess your test cases in Mathematica and export the used rule IDs. This is much easier and you can trace each integration problem. With this, you verify that using Rubi with Mathics follows the exact same steps through solving an integral.
3. We also covered the case you described above, i.e., what to do if a problem hangs or delivers the wrong result. I traced the Rubi call in Mathematica for these cases and exported each integration step. With this, you can verify that each transformation is correct.

You can [find here](https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-io/src/test/java/org/matheclipse/core/rubi/Exponentials.java#L1520-L1534) how this looks in Symja, although I seem to remember we only used 1-step integration tests in this case. When we ran into issues with an integration problem and dissected each Rubi expression, we created [a unit test like this](https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-io/src/test/java/org/matheclipse/core/rubi/issues/RubiIssue72.java).

I have two accompanying and undocumented Mathematica notebooks that I used back then, and I'm happy to share them and provide some help. One notebook is for creating unit tests that compare expressions or rule numbers on integration problems and exports a whole integration test suite. The other notebook shows how to trace Rubi calls and create unit tests for each evaluation happening during an integration.

Who do you think would be the best person to move this forward?

**Reply by rocky** [reply]
> Who do you think would be the best person to move this forward?

As happens with open-source projects, it is often unclear who is the "best" person.  Everybody has strengths and weaknesses, and everyone is volunteering as well.  So I can only speak for myself: I can't use a Mathematica notebook. (The project is not far enough along that we can read these in Mathics3).  And as is evident in this discussion, my knowledge of how WMA works is a bit limited.

A suggestion then is just to use this discussion; everyone who is interested, then will see.

**Reply by rocky** [reply]
> You use a test case like Int[F^(2 + 5*x), x] and compare your result with the result of Rubi. As @axkr already mentioned, this can be tricky because even in medium-sized problems, you run into issues where equivalent results cannot be verified with a simple string comparison.

Let's say `Int[F^(2 + 5*x), x` is put in a YAML file. Then we can list both what the WMA output is along with the Symja or Mathics3 output is when the the output is different. Then a simple string compare can be used.

**Comment by rocky** [comment]
@aravindh-krishnamoorthy I have created https://github.com/Mathics3/Mathics3-Rubi and have invited you as a maintainer.

I'd be grateful if you would work there instead of under the fork/branch in mathics core. Instead you can checkout this git repo as a [git submodule](https://git-scm.com/book/en/v2/Git-Tools-Submodules).

Thanks.
